### PR TITLE
Optimize the DISK State and reconstruct the state flow process

### PR DESCRIPTION
**Pull request by SSmallMonster**
## What this PR does / why we need it:
To simplify the management of disk states.

## What's changed
Disk states are classified into the following categories
* **Available**: represents the disk can be used which means:
            1. there is no filesystem or partitions exis
            2. the disk is not bound to any LocalDiskClaim object
* **Bound**: represents the disk is used already. There are follow-up use cases:
            1. used by system (e.g., rootfs)
            2. used by a LocalDiskClaim object
            3. there is already a filesystem or partition exist

### CODE REVIEW

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller.go [hunk 1]
```diff
@@ -112,29 +104,154 @@ func (r *ReconcileLocalDisk) Reconcile(_ context.Context, req reconcile.Request)
...
+		diskBound = false
+		err       error
+	)
+
+	if disk.Spec.ClaimRef != nil || disk.Spec.HasPartition {
+		diskBound = true
+	}
+
+	// Check if disk has been used and reserved at the same time
+	if diskBound && disk.Spec.Reserved {
```
Comment (sun7927) [inline pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:158]: This part should happen in processDiskBound()

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller.go [hunk 2]
```diff
@@ -112,29 +104,154 @@ func (r *ReconcileLocalDisk) Reconcile(_ context.Context, req reconcile.Request)
...
+func (r *ReconcileLocalDisk) processDiskBound(disk *v1alpha1.LocalDisk) error {
+	logCtx := log.Fields{"name": disk.Name}
+	log.WithFields(logCtx).Info("Start to processing Bound localdisk")
+
+	var (
+		err error
+	)
+
+	// Check if disk can be released
+	if disk.Spec.ClaimRef != nil && disk.Spec.HasPartition {
```
Comment (sun7927) [inline pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:244]: why can the disk be released when it has ClaimRef or HasPartition?

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller_test.go [hunk 3]
```diff
@@ -32,7 +32,7 @@ var (
 	vendorVMware                 = "VMware"
 	proSCSI                      = "scsi"
 	apiversion                   = "hwameistor.io/v1alpha1"
-	localDiskKind                = "LocalDisk"
+	localDiskKind                = "localDisk"
```
Comment (sun7927) [inline pkg/local-disk-manager/controller/localdisk/localdisk_controller_test.go:35]: why changed "LocalDisk" to "localDisk"?

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller.go [hunk 4]
```diff
@@ -112,29 +104,154 @@ func (r *ReconcileLocalDisk) Reconcile(_ context.Context, req reconcile.Request)
...
+	log.WithFields(logCtx).Info("Start to processing Available localdisk")
+
+	var (
+		diskBound = false
+		err       error
+	)
+
+	if disk.Spec.ClaimRef != nil || disk.Spec.HasPartition {
+		diskBound = true
+	}
```
[the remark this answers is not in the document]
Comment (SSmallMonster) [reply pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:155]: Disk bound can be triggered in two ways.
- One scenario is that a partition or file system is detected (which may be used by the system),
- The other scenario is that it is applied by the LDC. 

However, these two scenarios cannot occur at the same time. This is used to judge the occurrence of this exception and prevent users from applying for a disk with partitions.

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller.go (hunk 2 above)
Comment (SSmallMonster) [reply pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:244]: Seems mistake, it should use !HasPartition

File: pkg/local-disk-manager/controller/localdisk/localdisk_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sun7927) [reply pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:155]: in reconciling method, if a partition or filesystem is detected on a disk, the disk should transit to bound status immediately.

[the remark this answers is not in the document]
Comment (SSmallMonster) [reply pkg/local-disk-manager/controller/localdisk/localdisk_controller.go:154]: Solved~

**Comment by SSmallMonster** [comment]
@sun7927 Ready for review

File: deploy/crds/hwameistor.io_localdiskclaims_crd.yaml [hunk 5]
```diff
@@ -40,6 +40,11 @@ spec:
           spec:
             description: LocalDiskClaimSpec defines the desired state of LocalDiskClaim
             properties:
+              completed:
```
Comment (sun7927) [inline deploy/crds/hwameistor.io_localdiskclaims_crd.yaml:43]: why add "completed" into spec? should it be added into status? If it is added for disk expansion, "completed" may not be a good field name

Comment (SSmallMonster) [reply deploy/crds/hwameistor.io_localdiskclaims_crd.yaml:43]: This field is used for scenarios where LDC needs to be expanded. Users need a parameter to trigger the expansion behavior.

**Comment by sun7927** [comment]
LGTM

**Pull request merged**